src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/JunctionTreeKBestHaplotypeFinder.java

```diff
@@ -188,4 +228,201 @@ public boolean keepCycles() {
+                    .map(v -> junctionTreeLinkedDeBruinGraph.getJunctionTreeForNode((MultiDeBruijnVertex) v))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toList());
+            pathToAdd.markTreesAsVisited(treesPassed);
+            queue.add(pathToAdd);
+        }
+    }
+
+    //TODO this probably needs to be more testing...
+    //TODO thiss might be best computed in the paths as they are being expanded
+    private void annotatePathBasedOnGraph(final JTBestHaplotype<V, E> newPath, final JunctionTreeLinkedDeBruinGraph graph) {
+        int farthestReferenceEdgeReached = 0;
+        int numUpstreamRefEdgesEncountered = 0;
+        int lastReferenceEdgeVisited = 0;
+        for (E edge : newPath.getEdges()) {
+            List<Integer> refOccurances = ((MultiSampleEdge)edge).getReferencePathIndexes();
+
+            // if we are not on a reference edge don't worry
+            if (!refOccurances.isEmpty()) {
+                // find the next lowest ref occurance that is incrementally higher than our current one (assumes sorted ref occurrences list)
+                int refIndex = 0;
+                for (Integer index : refOccurances) {
+                    refIndex = index;
+                    if (refIndex > lastReferenceEdgeVisited) {
+                        break;
+                    }
+                }
+
+                // check if we are too far upstream
+                if (farthestReferenceEdgeReached > refIndex + DEFAULT_MAX_UPSTREAM_REFERENCE_JUMP_TO_ALLOW) {
+                    numUpstreamRefEdgesEncountered++;
+                } else if (refIndex > farthestReferenceEdgeReached) {
+                    farthestReferenceEdgeReached = refIndex;
+                }
+                lastReferenceEdgeVisited = refIndex;
+            }
+        }
+
+        // if we saw too many out of sequence reference edges then we report it as a potentially bad haplotype
+        if (numUpstreamRefEdgesEncountered > DEFAULT_NUM_UPSTREAM_REFERENCE_EDGES_TO_TOLERATE) {
+            newPath.setIsWonky(true);
+        }
+    }
+
+
+    /**
+     * Helper method that takes BestHaplotypePaths that may or may not start at a valid reference start position
+     *
+     * @return true if a valid reference-starting path was able to be constructed.
+     */
+    // TODO maybe if this fails to find we should include the reference path explicitly
+    private JTBestHaplotype<V, E> reconcilePathMissingReferenceStartPositions(JTBestHaplotype<V, E> pathToReconcile, List<JTBestHaplotype<V, E>> validReturnPaths, BaseGraph<V, E> graph) {
+        // check that the path is valid, if so don't modify it
+        if (  sources.contains(pathToReconcile.getVertices().get(0))) {
+            return pathToReconcile;
+        }
+
+        throw new RuntimeException("e");
+//
+//        V pivotalVerex = pathToReconcile.getFirstVertex();
```

Comment:
This was the old attempt to reconcile path heads after the fact. Something akin to this will be necessary in the future if we want to base head pasting on the similarity between ptaths. For now i think its okay to just remove it.

Reply:
Remove is good for now.